includes/class-fs-garbage-collector.php

```diff
@@ -0,0 +1,318 @@
+<?php
+    /**
+     * @package     Freemius
+     * @copyright   Copyright (c) 2015, Freemius, Inc.
+     * @license     https://www.gnu.org/licenses/gpl-3.0.html GNU General Public License Version 3
+     * @since       2.6.0
+     */
+
+    if ( ! defined( 'ABSPATH' ) ) {
+        exit;
+    }
+
+    class FS_Garbage_Collector {
+        #--------------------------------------------------------------------------------
+        #region Singleton
+        #--------------------------------------------------------------------------------
+
+        /**
+         * @var FS_Garbage_Collector
+         * @since 2.6.0
+         */
+        private static $_instance;
+
+        /**
+         * @param Freemius $freemius
+         *
+         * @return FS_Garbage_Collector
+         */
+        static function instance() {
+            if ( ! isset( self::$_instance ) ) {
+                self::$_instance = new self();
+            }
+
+            return self::$_instance;
+        }
+
+        #endregion
+
+        private function __construct() {
+        }
+
+        function init() {
+            require_once WP_FS__DIR_INCLUDES . '/class-fs-lock.php';
+
+            $lock = new FS_Lock( 'garbage_collection' );
+
+            /**
+             * Try to acquire lock for the next 60 sec based on the thread ID.
+             */
+            if ( ! $lock->try_lock( 60 ) ) {
+                return false;
+            }
+
+            $this->clean_up();
+
+            // Create a 1-day lock.
+            $lock->lock( WP_FS__TIME_24_HOURS_IN_SEC );
+        }
+
+        private function clean_up() {
+            $product_types = $this->get_product_types();

```

Comment:
@fajardoleo This function is a bit difficult and complex to go through and understand :)

I propose to refactor it like this


```php
private function clean_up() {
  $_accounts = FS_Options::instance( WP_FS__ACCOUNTS_OPTION_NAME, true );
  $products_to_delete = array();

  foreach ( $this->get_product_types() as $product_type ) {
    $products_to_delete = array_merge( 
       $products_to_delete, 
       $this->get_products_to_delete( $product_type, $_accounts ) 
    );
  }

  if ( ! empty( $products_to_delete ) ) {
    $this->delete_products( $products_to_delete, $_accounts );
    $this->delete_inactive_users( $_accounts );
  }
}

private function get_products_to_delete( $product_type, $accounts ) {
  $option_name      = ( $product_type . 's' );
  $all_product_data = $_accounts->get_option( $option_name, array() );

  $products_to_delete = array();

  foreach ( $all_product_data as $slug => $product_data ) {
    if ( $this->is_product_active( $slug, $product_data ) ) {
      continue;
    }

    // do your other stuff, but don't delete anything yet.
    // I think we can return just the `slug => $product_data` pair here.
  }

  return $products_to_delete;
}

private function delete_products( $products_to_delete, $_accounts ) {
  // Rearrange so that all add-ons are on top.

  // Delete the products + associated options. You can break this function down too.

  // Don't delete the users yet.
}

private function delete_inactive_users( $_accounts ) {
  // Loop and find all users.
  // Delete the users without any "install".
}
```

Another approach could be to break down this entire logic into multiple "strategy" classes. Statetiges would involve "how to delete inactive plugins", "how to delete inactive themes" etc. But I leave it to you, I don't think it is very important.

But please do see if/how we can simplify the big logic inside this function. It is very hard to understand and would be very difficult to debug. I would urge you to explore logic/algorithm where we would loop multiple times but the code would be readable and understandable.